metagraph/src/graph/alignment/dbg_aligner.hpp

```diff
@@ -24,377 +20,153 @@ class IDBGAligner {
 
     // Align the query sequence in the given orientation (false is forward,
     // true is reverse complement)
-    DBGQueryAlignment align_one_direction(const std::string_view query,
-                                          bool orientation) const;
+    void align_one_direction(DBGQueryAlignment &paths,
+                             bool orientation_to_align,
+                             const ISeeder<node_index> &seeder) const;
+
+    // Align both the forward and reverse complement of the query sequence,
+    // then report the best scoring alignment.
+    void align_best_direction(DBGQueryAlignment &paths,
+                              const ISeeder<node_index> &seeder,
+                              const ISeeder<node_index> &seeder_rc) const;
 
     // Align both forwards and backwards from a given seed. Procedure
     // 1. Given each seed, extend forward to produce an alignment A
     // 2. Reverse complement the alignment to get A', treated like a new seed
     // 3. Extend A' forwards
     // 4. Reverse complement A' to get the final alignment A''
-    DBGQueryAlignment align_both_directions(const std::string_view query) const;
-
-    // Align both the forward and reverse complement of the query sequence,
-    // then report the best scoring alignment.
-    DBGQueryAlignment align_forward_and_reverse_complement(const std::string_view query) const;
+    void align_both_directions(DBGQueryAlignment &paths,
+                               const ISeeder<node_index> &seeder) const;
 
     // Given alignments generated by a generator, add them to a priority queue
     // and add the top ones to paths.
-    virtual void align_aggregate(DBGQueryAlignment &paths,
-                                 const AlignmentGenerator &alignment_generator) const = 0;
-};
+    virtual void
+    align_aggregate(DBGQueryAlignment &paths,
+                    const AlignmentGenerator &alignment_generator) const;
 
+  protected:
+    const DeBruijnGraph &graph_;
+    DBGAlignerConfig config_;
+};
 
 template <class Seeder = ExactSeeder<>,
           class Extender = DefaultColumnExtender<>,
           class AlignmentCompare = std::less<Alignment<>>>
-class DBGAligner : public SeedAndExtendAligner<Seeder, Extender> {
+class DBGAligner : public ISeedAndExtendAligner<AlignmentCompare> {
   public:
-    typedef DeBruijnGraph::node_index node_index;
-    typedef Alignment<node_index> DBGAlignment;
-    typedef QueryAlignment<node_index> DBGQueryAlignment;
-    typedef typename DBGAlignment::score_t score_t;
+    typedef ISeedAndExtendAligner<AlignmentCompare> BaseAligner;
+    typedef IDBGAligner::node_index node_index;
+    typedef IDBGAligner::DBGAlignment DBGAlignment;
+    typedef IDBGAligner::DBGQueryAlignment DBGQueryAlignment;
+    typedef IDBGAligner::score_t score_t;
+    typedef IDBGAligner::QueryGenerator QueryGenerator;
+    typedef IDBGAligner::AlignmentCallback AlignmentCallback;
```

Comment:
For now it's like this, but #279 introduces another derived class called `LabeledDBGAligner` which inherits from `ISeedAndExtendAligner`, so that's why I structured it this way.

Reply:
Is it possible to inherit those new classes from this `SeedAndExtendAligner` (you would need to inherit `LabeledColumnExtender` from `IExtender`, which I think makes sense).

Reply:
Essentially, I want to have two different structures, which are currently 1) `ISeedAndExtendAligner` and 2) `DBGAligner`
1) Given a query, seeder, and extender, run the alignment engine and aggregate the results
2) Given a batch of queries, generate (seeder,extender) pairs and pass them to the alignment engine